### ISSUE TITLE
Green Deposit button remains disabled even when all required fields / sections are filled in

### ISSUE THREAD

**Issue by davidatmire** [DSpace/dspace-angular#1358]
**Describe the bug**
In some various situations which are not necessarily easy to reproduce, the green Deposit button at the end of the submission forms remains disabled even when all required fields / sections are filled in.

This prevents the submitter from actually depositing the item and completing the submission.

**To Reproduce**
It is not easy to provide a procedure which will result systematically in the problematic situation.

I have been able to reproduce it several times on demo7, with the following procedure :
1. New > item on demo7
2. Select "1-step workflow collection" (https://demo7.dspace.org/collections/51715dd3-5590-49f2-b227-6a663c849921
) in the modal
3. Fill the 2 required fields (Title and Date of issue), add a file and tick the licence box
4. The Deposit button is NOT enabled (while it should)
(5. After saving the item again, it gets enabled.)

We have noticed that the same issue appears in other scenarii such as:
- filling the form in a random order
- changing / adding / removing files
- filling repeatable mandatory fields

In order to provide you with a second/different procedure to reproduce the problem.
If you start from after step 5 of the above first procedure and:
6. Remove file
7. Remove the year of "Date of Issue"
8.  Select a value in the "Type" dropdown
9. Add file
10. Add year of "Date of issue" again
11. The Deposit button is NOT enabled (while it should)

There are several variations which will result in the same situation, i.e. the Deposit button not being enabled while it should be.
This situation is confusing for the submitter, who then thinks that he is not able to complete the submission of the item.

**Expected behavior**
Whether the Deposit button should be correctly enabled in ALL situations when all required fields / sections are filled in .

OR, another possible alternative would be for the Deposit button to always be enabled and return an error message if it is actioned while there are still missing required fields / sections.

**Comment by TanyaHolm** [comment]
UNSW is experiencing the same issue in our DSpace test instance: all mandatory metadata fields are populated, but the section displays orange warning - which means depositor cannot complete deposit.
It is hard to reproduce, but happens regularly.
We are going live on 1 December with submissions and this is a very worrying problem.

**Comment by tdonohue** [comment]
@davidatmire and @artlowel : While I can reproduce this issue, every time I reproduce it, it looks like the cause is that there are "Unsaved Changes".  So, here's what the buttons look like to me:
![unsaved-changes](https://user-images.githubusercontent.com/483997/140081226-3f300c32-29c2-4c53-8cc9-569baf989faf.png)

When I see that button combination, if I click "Save", then the "Deposit" button immediately enables.

I've also noticed that when this occurs, it's because one of the sections in the form still has an orange (or red) exclamation mark (!) notice.  As soon as all sections have green checkmarks, the deposit button enables.

This implies to me that the problem here is an auto-save problem in one or more of the sections.  Specifically, as I test this out, I think the problem is in the first "Describe" section (where you entire a Title and Date of Issue).  Even after entering required fields in that section, no save is ever triggered, and it still has an orange exclamation mark.  Even after uploading a file or checking the license, that first "Describe" section retains an orange exclamation mark.  It only turns into a green checkmark after you click the "Save" button.

While I'm not against the approach of enabling the "Deposit" button at all times, I'm not sure that's the proper fix if a _single section_ is causing the issues.  In this scenario, it looks (to me) like that first Describe section is simply never saving it's values or updating to a green checkbox until you manually click the "Save" button.

Glad to discuss this more as needed (perhaps even in a meeting).  I agree it needs to be fixed, but I'd like us to reanalyze the problem to see if it'd be easier to just fix by fixing the behavior of the first Describe section.

**Comment by Atmire-github** [comment]
@tdonohue : Thank you for looking into this.
I agree with your analysis and also think that this is an auto-save problem.
The route you want to take is also one I had initially envisioned in my first comment in this issue : 

> Whether the Deposit button should be correctly enabled in ALL situations when all required fields / sections are filled in .
> OR [...]

In the meanwhile , I have come to realise that there are technical difficulties associated with such a solution: It is my understanding that previous issues have led to reducing the rate of autosaving while this issue would rather result in increasing it. Indeed, if you want an autosave to be triggered for the scenario provided in my initial comment, you would likely need to trigger it for each metadata field (after each field has been filled in). Especially, for the second part of my test scenario, it would even be possible that the user does not leave the field itself when trying to finish his submission (trying to click the green deposit button); such a situation would make it even more complicated to have a specific event which would trigger the autosave.

Hence, even if I agree that we should indeed try to fix/improve autosave, I think that, right now, always enabling the deposit button would be a significant step forward.

Furthermore, it is worth mentioning that users of DSpace 6 repositories (or previous versions) are not used to the "Save" button. They do have a "Save & Exit" button, but not a "Save" button. The presence of these 2 buttons can be a bit confusing for these users. Having the green "Deposit" button always enabled, would most likely make it easier for them to complete their submission (I am personally not even sure that keeping the "Save" button would then have any added value, but this is more of a personal opinion).

What do you think ?

**Comment by davidatmire** [comment]
@tdonohue : Thank you for looking into this.

 I agree with your analysis and also think that this is an auto-save problem. The route you want to take is also one I had initially envisioned in my first comment in this issue :
> Whether the Deposit button should be correctly enabled in ALL situations when all required fields / sections are filled in .
>  OR [...]

In the meanwhile , I have come to realise that there are technical difficulties associated with such a solution: It is my understanding that previous issues have led to reducing the rate of autosaving while this issue would rather result in increasing it. Indeed, if you want an autosave to be triggered for the scenario provided in my initial comment, you would likely need to trigger it for each metadata field (after each field has been filled in). Especially, for the second part of my test scenario, it would even be possible that the user does not leave the field itself when trying to finish his submission (trying to click the green deposit button); such a situation would make it even more complicated to have a specific event which would trigger the autosave.

Hence, even if I agree that we should indeed try to fix/improve autosave, I think that, right now, always enabling the deposit button would be a significant step forward.

Furthermore, it is worth mentioning that users of DSpace 6 repositories (or previous versions) are not used to the "Save" button. They do have a "Save & Exit" button, but not a "Save" button. The presence of these 2 buttons can be a bit confusing for these users. Having the green "Deposit" button always enabled, would most likely make it easier for them to complete their submission (I am personally not even sure that keeping the "Save" button would then have any added value, but this is more of a personal opinion).

What do you think ?

**Comment by tdonohue** [comment]
@davidatmire : Thanks for your additional thoughts here & you bring up some good points about the older (v6 and below) behavior versus the new behavior & general usability.

I'm OK with your & @artlowel 's idea to simply enable the Deposit button at all times, provided that we can ensure error handling, etc is still working appropriately for required fields/sections.  I'd recommend we do keep the "Save" button for now, as I think it provides an easy way to remove that "Unsaved changes" notification -- and I feel that "Unsaved changes" message is necessary for usability/clarity.

One minor recommendation would be that clicking the "Deposit" button when "Unsaved changes" exist should somehow make it clear that it is *first saving the changes* and then attempting deposit (otherwise users may worry that the unsaved changes are not part of the deposit).  Even simply changing that "Unsaved changes" message to "Saving and depositing..." after "Deposit" is clicked would be a good improvement.

In any case, I'm OK with your idea/approach, provided that we ensure it provides improved usability to the form (and I think it can do so, provided error handling and saving before depositing are both clearly handled).

**Comment by tdonohue** [comment]
@artlowel : Per today's meeting, adding your initial estimate of 6hrs & assigning back to you to delegate to your team for 7.2.  Your and @davidatmire 's design is fine per agreement in our meeting.
